File: src/wwwroot/js/telnyxwebrtcinterop.js
export class TelnyxWebRtcInterop {
    constructor() {
        this._clients = new Map();
        this._observers = new Map();

        // one‑time window unload cleanup
        if (typeof window !== 'undefined' && !window._telnyxUnloadHandlerAdded) {
            window.addEventListener('unload', () => {
                for (const id of this._clients.keys()) this.unmount(id);
            });
            window._telnyxUnloadHandlerAdded = true;
        }
    }

    create( Id!, optionsJson, dotNetCallback) {
        if (this._clients.has(elementId)) {
            console.warn(`Telnyx WebRTC client already exists for "${elementId}".`);
            return;
        }

        const config = JSON.parse(optionsJson);

        const wrapper = {
            id: Id!,
            dotNetCallback,
            reconnectCount: 0,
            config,
            client: null
        };

        this._clients.set( Id!, wrapper);
        this._createClient(wrapper);
        dotNetCallback.invokeMethodAsync('HandleTelnyxEvent', 'initialized', '');
    }

    _createClient(wrapper) {
        const { config } = wrapper;

        wrapper.client = new TelnyxWebRTC.TelnyxRTC(config.initOptions);

        wrapper.client.remoteElement = config.initOptions.remoteElement;
        wrapper.client.localElement = config.initOptions.localElement;

        config.audio ? wrapper.client.enableMicrophone() : wrapper.client.disableMicrophone();
        config.video ? wrapper.client.enableWebcam() : wrapper.client.disableWebcam();

        this._bindEvents(wrapper);
        wrapper.client.connect();
        
        // --- Forward all raw WebSocket messages to .NET ---
        // Find the underlying WebSocket and hook into onmessage
        const ws = wrapper.client.connection && wrapper.client.connection._wsClient;
        if (ws) {
            ws.addEventListener('message', (evt) => {
                if (wrapper.dotNetCallback && evt && evt.data) {
                    wrapper.dotNetCallback.invokeMethodAsync('HandleTelnyxEvent', 'rawSocketMessage', evt.data);
                }
            });
        }
    }

    /* ---------- EVENT BINDING & HANDLERS ---------- */

    _bindEvents(wrapper) {
        const { client, dotNetCallback } = wrapper;

        const safeStringify = (o, indent = 2) => {
            const seen = new WeakSet();
            return JSON.stringify(o, (k, v) => {
                if (typeof v === 'object' && v !== null) {
                    if (seen.has(v)) return '[Circular]';
                    seen.add(v);
                }
                return v;
            }, indent);
        };
        const safeInvoke = (evt, args) =>
            dotNetCallback.invokeMethodAsync('HandleTelnyxEvent', evt, safeStringify(args));

        // session‑level events we simply forward
        [
            'ready', 'error', 'socket.open', 'socket.close',
            'socket.error', 'reconnecting', 'reconnected', 'disconnected',
            'socket.message', 'ping', 'pong'
        ].forEach(e => client.on(`telnyx.${e}`, arg => safeInvoke(e, arg)));

        // central handler for call updates & inbound ringing
        client.on('telnyx.notification', notification => {
            safeInvoke('notification', notification);

            if (notification.type !== 'callUpdate' || !notification.call) return;

            this._handleCallUpdate(wrapper, notification.call, safeInvoke);
        });
    }

    /**
     * Handles all call state transitions (both outbound & inbound).
     */
    _handleCallUpdate(wrapper, call, safeInvoke) {
        const { config } = wrapper;
        const state = call.state;

        // Always keep the reference fresh
        wrapper.client.currentCall = call;

        switch (state) {
            case 'ringing': {                        // <-- INCOMING
                safeInvoke('incoming', call);

                if (config.autoAnswer) {
                    try { call.answer(); }
                    catch (err) { console.error('Auto‑answer failed:', err); }
                }
                break;
            }

            case 'active':                           // answered / connected
                safeInvoke('incomingAnswered', call);
                wrapper.reconnectCount = 0;
                break;

            case 'done':
            case 'disconnected':                    // call ended (abnormal or connection loss)
                safeInvoke('incomingRejected', call);
                wrapper.client.currentCall = null;

                if (config.autoReconnect &&
                    wrapper.reconnectCount < config.reconnectAttempts) {
                    wrapper.reconnectCount++;
                    setTimeout(() => this._createClient(wrapper), config.reconnectDelay);
                }
                break;

            case 'hangup':                          // explicit/intentional end
            case 'destroy':                      
            case 'purge':                         
                safeInvoke('incomingRejected', call);
                wrapper.client.currentCall = null;
                // Do NOT auto-reconnect here
                break;

            default:
                // other states: new, held, hold, etc. – still forward
                safeInvoke(`state:${state}`, call);
        }
    }


    call( Id!, optionsJson) {
        const wrapper = this._clients.get(elementId);
        if (!wrapper?.client) return;

        // Prevent new call if one is already active
        if (wrapper.client.currentCall) {
            console.warn('TelnyxWebRtcInterop: Tried to start a new call while one is already active. Hang up first.');
            return;
        }

        const callOptions = JSON.parse(optionsJson);
        let call = null;
        try {
            call = wrapper.client.newCall(callOptions);
        } catch (err) {
            console.error('TelnyxWebRtcInterop: Exception during newCall:', err);
        }
        if (call) {
            wrapper.client.currentCall = call;
        } else {
            console.warn('TelnyxWebRtcInterop: Failed to create a new call.');
        }
    }

    answer( Id!, optionsJson) {
        const wrapper = this._clients.get(elementId);
        if (!wrapper?.client?.currentCall) {
            console.warn('TelnyxWebRtcInterop: Tried to answer but no currentCall is set.');
            return;
        }

        if (optionsJson) {
            const options = JSON.parse(optionsJson);
            wrapper.client.currentCall.answer(options);
        } else {
            wrapper.client.currentCall.answer();
        }
    }

    hangup( Id!, optionsJson) {
        const wrapper = this._clients.get(elementId);
        if (!wrapper?.client?.currentCall) {
            console.warn('TelnyxWebRtcInterop: Tried to hangup but no currentCall is set.');
            return;
        }

        if (optionsJson) {
            const options = JSON.parse(optionsJson);
            wrapper.client.currentCall.hangup(options);
        } else {
            wrapper.client.currentCall.hangup();
        }
        // Clean up currentCall after hangup
        wrapper.client.currentCall = null;
    }

    muteAudio(elementId) {
        this._clients.get(elementId)?.client?.currentCall?.muteAudio();
    }

    unmuteAudio(elementId) {
        this._clients.get(elementId)?.client?.currentCall?.unmuteAudio();
    }

    toggleAudioMute(elementId) {
        this._clients.get(elementId)?.client?.currentCall?.toggleAudioMute();
    }

    muteVideo(elementId) {
        this._clients.get(elementId)?.client?.currentCall?.muteVideo();
    }

    unmuteVideo(elementId) {
        this._clients.get(elementId)?.client?.currentCall?.unmuteVideo();
    }

    toggleVideoMute(elementId) {
        this._clients.get(elementId)?.client?.currentCall?.toggleVideoMute();
    }

    deaf(elementId) {
        this._clients.get(elementId)?.client?.currentCall?.deaf();
    }

    undeaf(elementId) {
        this._clients.get(elementId)?.client?.currentCall?.undeaf();
    }

    toggleDeaf(elementId) {
        this._clients.get(elementId)?.client?.currentCall?.toggleDeaf();
    }

    hold(elementId) {
        this._clients.get(elementId)?.client?.currentCall?.hold();
    }

    unhold(elementId) {
        this._clients.get(elementId)?.client?.currentCall?.unhold();
    }

    toggleHold(elementId) {
        this._clients.get(elementId)?.client?.currentCall?.toggleHold();
    }

    dtmf( Id!, digit) {
        this._clients.get(elementId)?.client?.currentCall?.dtmf(digit);
    }

    message( Id!, to, body) {
        this._clients.get(elementId)?.client?.currentCall?.message(to, body);
    }

    setAudioInDevice( Id!, deviceId) {
        this._clients.get(elementId)?.client?.currentCall?.setAudioInDevice(deviceId);
    }

    setVideoDevice( Id!, deviceId) {
        this._clients.get(elementId)?.client?.currentCall?.setVideoDevice(deviceId);
    }

    setAudioOutDevice( Id!, deviceId) {
        this._clients.get(elementId)?.client?.currentCall?.setAudioOutDevice(deviceId);
    }

    startScreenShare( Id!, optionsJson) {
        const wrapper = this._clients.get(elementId);
        if (!wrapper?.client?.currentCall) return;

        const options = optionsJson ? JSON.parse(optionsJson) : {};
        wrapper.client.currentCall.startScreenShare(options);
    }

    stopScreenShare(elementId) {
        this._clients.get(elementId)?.client?.currentCall?.stopScreenShare();
    }

    setAudioBandwidth( Id!, bps) {
        this._clients.get(elementId)?.client?.currentCall?.setAudioBandwidthEncodingsMaxBps(bps);
    }

    setVideoBandwidth( Id!, bps) {
        this._clients.get(elementId)?.client?.currentCall?.setVideoBandwidthEncodingsMaxBps(bps);
    }

    getDevices(elementId) {
        return new Promise((resolve) => {
            const client = this._clients.get(elementId)?.client;
            if (!client) {
                resolve(JSON.stringify([]));
                return;
            }

            client.getDevices().then(devices => {
                resolve(JSON.stringify(devices));
            }).catch(() => {
                resolve(JSON.stringify([]));
            });
        });
    }

    getVideoDevices(elementId) {
        return new Promise((resolve) => {
            const client = this._clients.get(elementId)?.client;
            if (!client) {
                resolve(JSON.stringify([]));
                return;
            }

            client.getVideoDevices().then(devices => {
                resolve(JSON.stringify(devices));
            }).catch(() => {
                resolve(JSON.stringify([]));
            });
        });
    }

    getAudioInDevices(elementId) {
        return new Promise((resolve) => {
            const client = this._clients.get(elementId)?.client;
            if (!client) {
                resolve(JSON.stringify([]));
                return;
            }

            client.getAudioInDevices().then(devices => {
                resolve(JSON.stringify(devices));
            }).catch(() => {
                resolve(JSON.stringify([]));
            });
        });
    }

    getAudioOutDevices(elementId) {
        return new Promise((resolve) => {
            const client = this._clients.get(elementId)?.client;
            if (!client) {
                resolve(JSON.stringify([]));
                return;
            }

            client.getAudioOutDevices().then(devices => {
                resolve(JSON.stringify(devices));
            }).catch(() => {
                resolve(JSON.stringify([]));
            });
        });
    }

    checkPermissions( Id!, audio = true, video = true) {
        return new Promise((resolve) => {
            const client = this._clients.get(elementId)?.client;
            if (!client) {
                resolve(false);
                return;
            }

            client.checkPermissions(audio, video).then(result => {
                resolve(result);
            }).catch(() => {
                resolve(false);
            });
        });
    }

    setAudioSettings( Id!, settingsJson) {
        return new Promise((resolve) => {
            const client = this._clients.get(elementId)?.client;
            if (!client) {
                resolve(false);
                return;
            }

            const settings = JSON.parse(settingsJson);
            client.setAudioSettings(settings).then(() => {
                resolve(true);
            }).catch(() => {
                resolve(false);
            });
        });
    }

    setVideoSettings( Id!, settingsJson) {
        return new Promise((resolve) => {
            const client = this._clients.get(elementId)?.client;
            if (!client) {
                resolve(false);
                return;
            }

            const settings = JSON.parse(settingsJson);
            client.setVideoSettings(settings).then(() => {
                resolve(true);
            }).catch(() => {
                resolve(false);
            });
        });
    }

    enableMicrophone(elementId) {
        this._clients.get(elementId)?.client?.enableMicrophone();
    }

    disableMicrophone(elementId) {
        this._clients.get(elementId)?.client?.disableMicrophone();
    }

    enableWebcam(elementId) {
        this._clients.get(elementId)?.client?.enableWebcam();
    }

    disableWebcam(elementId) {
        this._clients.get(elementId)?.client?.disableWebcam();
    }

    toggleAudio( Id!, enabled) {
        this._clients.get(elementId)?.client?.localStream?.getAudioTracks()?.forEach(t => t.enabled = enabled);
    }

    toggleVideo( Id!, enabled) {
        this._clients.get(elementId)?.client?.localStream?.getVideoTracks()?.forEach(t => t.enabled = enabled);
    }

    setAudioVolume( Id!, volume) {
        const el = document.getElementById(elementId);
        if (el && el.tagName === "AUDIO") {
            el.volume = Math.max(0, Math.min(1, volume));
        }
    }

    disconnect(elementId) {
        const wrapper = this._clients.get(elementId);
        if (!wrapper?.client) return;

        wrapper.client.off('telnyx.ready');
        wrapper.client.off('telnyx.notification');
        wrapper.client.off('telnyx.error');
        wrapper.client.off('telnyx.socket.close');

        // Clean up currentCall on disconnect
        if (wrapper.client.currentCall) {
            wrapper.client.currentCall = null;
        }

        wrapper.client.disconnect();
    }

    reconnect(elementId) {
        const wrapper = this._clients.get(elementId);
        if (!wrapper) return;

        this.disconnect(elementId);
        wrapper.reconnectCount = 0;
        // Clean up observer before re-creating client
        this._disconnectObserver(elementId);
        this._createClient(wrapper);
    }

    connect(elementId) {
        const wrapper = this._clients.get(elementId);
        if (wrapper?.client) {
            wrapper.client.connect();
        }
    }

    unmount(elementId) {
        const wrapper = this._clients.get(elementId);
        if (!wrapper) return;

        try {
            this.disconnect(elementId);
        } catch (e) {
            console.warn(`Error during disconnect of client "${elementId}":`, e);
        }

        this._clients.delete(elementId);
        this._disconnectObserver(elementId);
    }

    createObserver(elementId) {
        const el = document.getElementById(elementId);
        if (!el || !el.parentNode) {
            console.warn(`Element "${elementId}" not found for mutation observer.`);
            return;
        }

        const observer = new MutationObserver((mutations) => {
            const removed = mutations.some(m =>
                Array.from(m.removedNodes).includes(el)
            );

            if (removed) {
                this.unmount(elementId);
            }
        });

        observer.observe(el.parentNode, { childList: true });
        this._observers.set( Id!, observer);
    }

    _disconnectObserver(elementId) {
        const observer = this._observers.get(elementId);
        if (observer) {
            observer.disconnect();
            this._observers.delete(elementId);
        }
    }

    // Conference control methods
    listVideoLayouts(elementId) {
        this._clients.get(elementId)?.client?.currentCall?.listVideoLayouts();
    }

    setVideoLayout( Id!, layout, canvas) {
        this._clients.get(elementId)?.client?.currentCall?.setVideoLayout(layout, canvas);
    }

    playMedia( Id!, source) {
        this._clients.get(elementId)?.client?.currentCall?.playMedia(source);
    }

    stopMedia(elementId) {
        this._clients.get(elementId)?.client?.currentCall?.stopMedia();
    }

    startRecord( Id!, filename) {
        this._clients.get(elementId)?.client?.currentCall?.startRecord(filename);
    }

    stopRecord(elementId) {
        this._clients.get(elementId)?.client?.currentCall?.stopRecord();
    }

    sendChatMessage( Id!, message, type) {
        this._clients.get(elementId)?.client?.currentCall?.sendChatMessage(message, type);
    }

    snapshot( Id!, filename) {
        this._clients.get(elementId)?.client?.currentCall?.snapshot(filename);
    }

    muteMic( Id!, participantId) {
        this._clients.get(elementId)?.client?.currentCall?.muteMic(participantId);
    }

    muteVideoParticipant( Id!, participantId) {
        this._clients.get(elementId)?.client?.currentCall?.muteVideo(participantId);
    }

    kick( Id!, participantId) {
        this._clients.get(elementId)?.client?.currentCall?.kick(participantId);
    }

    volumeUp( Id!, participantId) {
        this._clients.get(elementId)?.client?.currentCall?.volumeUp(participantId);
    }

    volumeDown( Id!, participantId) {
        this._clients.get(elementId)?.client?.currentCall?.volumeDown(participantId);
    }

    getCallStats(elementId) {
        return new Promise((resolve) => {
            const call = this._clients.get(elementId)?.client?.currentCall;
            if (!call) {
                resolve(null);
                return;
            }

            const stats = [];
            call.getStats((stat) => {
                stats.push(stat);
                resolve(JSON.stringify(stats));
            });
        });
    }
}

window.TelnyxWebRtcInterop = new TelnyxWebRtcInterop();

File: src/TelnyxWebRtcInterop.cs
using Microsoft.JSInterop;
using Soenneker.Blazor.Utils.ResourceLoader.Abstract;
using Soenneker.Extensions.ValueTask;
using Soenneker.Telnyx.Blazor.WebRtc.Abstract;
using Soenneker.Utils.Json;
using System.Threading;
using System.Threading.Tasks;
using Soenneker.Asyncs.Initializers;
using Soenneker.Telnyx.Blazor.WebRtc.Configuration;

namespace Soenneker.Telnyx.Blazor.WebRtc;

///<inheritdoc cref="ITelnyxWebRtcInterop"/>
public sealed class TelnyxWebRtcInterop : ITelnyxWebRtcInterop
{
    private readonly IJSRuntime _jsRuntime;
    private readonly IResourceLoader _resourceLoader;

    private readonly AsyncInitializer<bool> _scriptInitializer;

    private const string _module = "Soenneker.Telnyx.Blazor.WebRtc/js/telnyxwebrtcinterop.js";
    private const string _moduleName = "TelnyxWebRtcInterop";

    public TelnyxWebRtcInterop(IJSRuntime jsRuntime, IResourceLoader resourceLoader)
    {
        _jsRuntime = jsRuntime;
        _resourceLoader = resourceLoader;

        _scriptInitializer = new AsyncInitializer<bool>(InitializeScripts);
    }

    private async ValueTask InitializeScripts(bool useCdn, CancellationToken token)
    {
        if (useCdn)
        {
            await _resourceLoader.LoadScriptAndWaitForVariable("https://cdn.jsdelivr.net/npm/@telnyx/webrtc@2.22.17/lib/bundle.js", "TelnyxWebRTC",
                integrity: "sha256-uiKtParibFFpEaHhD+X8rgPhdUAWgcDhHKXwTzqARbE=", cancellationToken: token);
        }
        else
        {
            await _resourceLoader.LoadScriptAndWaitForVariable("_content/Soenneker.Telnyx.Blazor.WebRtc/js/telnyxwebrtc.js", "TelnyxWebRTC",
                cancellationToken: token);
        }

        await _resourceLoader.ImportModuleAndWaitUntilAvailable(_module, _moduleName, 100, token);
    }

    public ValueTask Initialize(bool useCdn = true, CancellationToken cancellationToken = default)
    {
        return _scriptInitializer.Init(useCdn, cancellationToken);
    }

    public async ValueTask Create(string id, DotNetObjectReference<TelnyxWebRtc> dotNetObjectRef, TelnyxClientOptions options,
        CancellationToken cancellationToken = default)
    {
        await _scriptInitializer.Init(true, cancellationToken)
                                .NoSync();
        string? json = JsonUtil.Serialize(options);

        await _jsRuntime.InvokeVoidAsync("TelnyxWebRtcInterop.create", cancellationToken, id, json, dotNetObjectRef)
                        .NoSync();
    }

    public ValueTask CreateObserver(string id, CancellationToken cancellationToken = default)
    {
        return _jsRuntime.InvokeVoidAsync("TelnyxWebRtcInterop.createObserver", cancellationToken, id);
    }

    public ValueTask Call(string id, TelnyxCallOptions callOptions, CancellationToken cancellationToken = default) =>
        _jsRuntime.InvokeVoidAsync("TelnyxWebRtcInterop.call", cancellationToken, id, JsonUtil.Serialize(callOptions));

    public ValueTask Answer(string id, TelnyxAnswerOptions? options = null, CancellationToken cancellationToken = default)
    {
        if (options != null)
            return _jsRuntime.InvokeVoidAsync("TelnyxWebRtcInterop.answer", cancellationToken, id, JsonUtil.Serialize(options));

        return _jsRuntime.InvokeVoidAsync("TelnyxWebRtcInterop.answer", cancellationToken, id);
    }

    public ValueTask Hangup(string id, TelnyxHangupOptions? options = null, CancellationToken cancellationToken = default)
    {
        if (options != null)
            return _jsRuntime.InvokeVoidAsync("TelnyxWebRtcInterop.hangup", cancellationToken, id, JsonUtil.Serialize(options));

        return _jsRuntime.InvokeVoidAsync("TelnyxWebRtcInterop.hangup", cancellationToken, id);
    }

    public ValueTask MuteAudio(string id, CancellationToken cancellationToken = default) =>
        _jsRuntime.InvokeVoidAsync("TelnyxWebRtcInterop.muteAudio", cancellationToken, id);

    public ValueTask UnmuteAudio(string id, CancellationToken cancellationToken = default) =>
        _jsRuntime.InvokeVoidAsync("TelnyxWebRtcInterop.unmuteAudio", cancellationToken, id);

    public ValueTask ToggleAudioMute(string id, CancellationToken cancellationToken = default) =>
        _jsRuntime.InvokeVoidAsync("TelnyxWebRtcInterop.toggleAudioMute", cancellationToken, id);

    public ValueTask MuteVideo(string id, CancellationToken cancellationToken = default) =>
        _jsRuntime.InvokeVoidAsync("TelnyxWebRtcInterop.muteVideo", cancellationToken, id);

    public ValueTask UnmuteVideo(string id, CancellationToken cancellationToken = default) =>
        _jsRuntime.InvokeVoidAsync("TelnyxWebRtcInterop.unmuteVideo", cancellationToken, id);

    public ValueTask ToggleVideoMute(string id, CancellationToken cancellationToken = default) =>
        _jsRuntime.InvokeVoidAsync("TelnyxWebRtcInterop.toggleVideoMute", cancellationToken, id);

    public ValueTask Deaf(string id, CancellationToken cancellationToken = default) =>
        _jsRuntime.InvokeVoidAsync("TelnyxWebRtcInterop.deaf", cancellationToken, id);

    public ValueTask Undeaf(string id, CancellationToken cancellationToken = default) =>
        _jsRuntime.InvokeVoidAsync("TelnyxWebRtcInterop.undeaf", cancellationToken, id);

    public ValueTask ToggleDeaf(string id, CancellationToken cancellationToken = default) =>
        _jsRuntime.InvokeVoidAsync("TelnyxWebRtcInterop.toggleDeaf", cancellationToken, id);

    public ValueTask Hold(string id, CancellationToken cancellationToken = default) =>
        _jsRuntime.InvokeVoidAsync("TelnyxWebRtcInterop.hold", cancellationToken, id);

    public ValueTask Unhold(string id, CancellationToken cancellationToken = default) =>
        _jsRuntime.InvokeVoidAsync("TelnyxWebRtcInterop.unhold", cancellationToken, id);

    public ValueTask ToggleHold(string id, CancellationToken cancellationToken = default) =>
        _jsRuntime.InvokeVoidAsync("TelnyxWebRtcInterop.toggleHold", cancellationToken, id);

    public ValueTask Dtmf(string id, string digit, CancellationToken cancellationToken = default) =>
        _jsRuntime.InvokeVoidAsync("TelnyxWebRtcInterop.dtmf", cancellationToken, id, digit);

    public ValueTask Message(string id, string to, string body, CancellationToken cancellationToken = default) =>
        _jsRuntime.InvokeVoidAsync("TelnyxWebRtcInterop.message", cancellationToken, id, to, body);

    public ValueTask SetAudioInDevice(string id, string deviceId, CancellationToken cancellationToken = default) =>
        _jsRuntime.InvokeVoidAsync("TelnyxWebRtcInterop.setAudioInDevice", cancellationToken, id, deviceId);

    public ValueTask SetVideoDevice(string id, string deviceId, CancellationToken cancellationToken = default) =>
        _jsRuntime.InvokeVoidAsync("TelnyxWebRtcInterop.setVideoDevice", cancellationToken, id, deviceId);

    public ValueTask SetAudioOutDevice(string id, string deviceId, CancellationToken cancellationToken = default) =>
        _jsRuntime.InvokeVoidAsync("TelnyxWebRtcInterop.setAudioOutDevice", cancellationToken, id, deviceId);

    public ValueTask StartScreenShare(string id, TelnyxScreenShareOptions? options = null, CancellationToken cancellationToken = default)
    {
        if (options != null)
            return _jsRuntime.InvokeVoidAsync("TelnyxWebRtcInterop.startScreenShare", cancellationToken, id, JsonUtil.Serialize(options));

        return _jsRuntime.InvokeVoidAsync("TelnyxWebRtcInterop.startScreenShare", cancellationToken, id);
    }

    public ValueTask StopScreenShare(string id, CancellationToken cancellationToken = default) =>
        _jsRuntime.InvokeVoidAsync("TelnyxWebRtcInterop.stopScreenShare", cancellationToken, id);

    public ValueTask SetAudioBandwidth(string id, int bps, CancellationToken cancellationToken = default) =>
        _jsRuntime.InvokeVoidAsync("TelnyxWebRtcInterop.setAudioBandwidth", cancellationToken, id, bps);

    public ValueTask SetVideoBandwidth(string id, int bps, CancellationToken cancellationToken = default) =>
        _jsRuntime.InvokeVoidAsync("TelnyxWebRtcInterop.setVideoBandwidth", cancellationToken, id, bps);

    public ValueTask<string> GetDevices(string id, CancellationToken cancellationToken = default) =>
        _jsRuntime.InvokeAsync<string>("TelnyxWebRtcInterop.getDevices", cancellationToken, id);

    public ValueTask<string> GetVideoDevices(string id, CancellationToken cancellationToken = default) =>
        _jsRuntime.InvokeAsync<string>("TelnyxWebRtcInterop.getVideoDevices", cancellationToken, id);

    public ValueTask<string> GetAudioInDevices(string id, CancellationToken cancellationToken = default) =>
        _jsRuntime.InvokeAsync<string>("TelnyxWebRtcInterop.getAudioInDevices", cancellationToken, id);

    public ValueTask<string> GetAudioOutDevices(string id, CancellationToken cancellationToken = default) =>
        _jsRuntime.InvokeAsync<string>("TelnyxWebRtcInterop.getAudioOutDevices", cancellationToken, id);

    public ValueTask<bool> CheckPermissions(string id, bool audio = true, bool video = true, CancellationToken cancellationToken = default) =>
        _jsRuntime.InvokeAsync<bool>("TelnyxWebRtcInterop.checkPermissions", cancellationToken, id, audio, video);

    public ValueTask<bool> SetAudioSettings(string id, TelnyxAudioSettings settings, CancellationToken cancellationToken = default) =>
        _jsRuntime.InvokeAsync<bool>("TelnyxWebRtcInterop.setAudioSettings", cancellationToken, id, JsonUtil.Serialize(settings));

    public ValueTask<bool> SetVideoSettings(string id, TelnyxVideoSettings settings, CancellationToken cancellationToken = default) =>
        _jsRuntime.InvokeAsync<bool>("TelnyxWebRtcInterop.setVideoSettings", cancellationToken, id, JsonUtil.Serialize(settings));

    public ValueTask EnableMicrophone(string id, CancellationToken cancellationToken = default) =>
        _jsRuntime.InvokeVoidAsync("TelnyxWebRtcInterop.enableMicrophone", cancellationToken, id);

    public ValueTask DisableMicrophone(string id, CancellationToken cancellationToken = default) =>
        _jsRuntime.InvokeVoidAsync("TelnyxWebRtcInterop.disableMicrophone", cancellationToken, id);

    public ValueTask EnableWebcam(string id, CancellationToken cancellationToken = default) =>
        _jsRuntime.InvokeVoidAsync("TelnyxWebRtcInterop.enableWebcam", cancellationToken, id);

    public ValueTask DisableWebcam(string id, CancellationToken cancellationToken = default) =>
        _jsRuntime.InvokeVoidAsync("TelnyxWebRtcInterop.disableWebcam", cancellationToken, id);

    public ValueTask ToggleAudio(string id, bool enabled, CancellationToken cancellationToken = default) =>
        _jsRuntime.InvokeVoidAsync("TelnyxWebRtcInterop.toggleAudio", cancellationToken, id, enabled);

    public ValueTask ToggleVideo(string id, bool enabled, CancellationToken cancellationToken = default) =>
        _jsRuntime.InvokeVoidAsync("TelnyxWebRtcInterop.toggleVideo", cancellationToken, id, enabled);

    public ValueTask Disconnect(string id, CancellationToken cancellationToken = default) =>
        _jsRuntime.InvokeVoidAsync("TelnyxWebRtcInterop.disconnect", cancellationToken, id);

    public ValueTask Reconnect(string id, CancellationToken cancellationToken = default) =>
        _jsRuntime.InvokeVoidAsync("TelnyxWebRtcInterop.reconnect", cancellationToken, id);

    public ValueTask Unmount(string id, CancellationToken cancellationToken = default) =>
        _jsRuntime.InvokeVoidAsync("TelnyxWebRtcInterop.unmount", cancellationToken, id);

    public ValueTask Connect(string id, CancellationToken cancellationToken = default) =>
        _jsRuntime.InvokeVoidAsync("TelnyxWebRtcInterop.connect", cancellationToken, id);

    // Conference control methods
    public ValueTask ListVideoLayouts(string id, CancellationToken cancellationToken = default) =>
        _jsRuntime.InvokeVoidAsync("TelnyxWebRtcInterop.listVideoLayouts", cancellationToken, id);

    public ValueTask SetVideoLayout(string id, string layout, string? canvas = null, CancellationToken cancellationToken = default) =>
        _jsRuntime.InvokeVoidAsync("TelnyxWebRtcInterop.setVideoLayout", cancellationToken, id, layout, canvas);

    public ValueTask PlayMedia(string id, string source, CancellationToken cancellationToken = default) =>
        _jsRuntime.InvokeVoidAsync("TelnyxWebRtcInterop.playMedia", cancellationToken, id, source);

    public ValueTask StopMedia(string id, CancellationToken cancellationToken = default) =>
        _jsRuntime.InvokeVoidAsync("TelnyxWebRtcInterop.stopMedia", cancellationToken, id);

    public ValueTask StartRecord(string id, string filename, CancellationToken cancellationToken = default) =>
        _jsRuntime.InvokeVoidAsync("TelnyxWebRtcInterop.startRecord", cancellationToken, id, filename);

    public ValueTask StopRecord(string id, CancellationToken cancellationToken = default) =>
        _jsRuntime.InvokeVoidAsync("TelnyxWebRtcInterop.stopRecord", cancellationToken, id);

    public ValueTask SendChatMessage(string id, string message, string? type = null, CancellationToken cancellationToken = default) =>
        _jsRuntime.InvokeVoidAsync("TelnyxWebRtcInterop.sendChatMessage", cancellationToken, id, message, type);

    public ValueTask Snapshot(string id, string filename, CancellationToken cancellationToken = default) =>
        _jsRuntime.InvokeVoidAsync("TelnyxWebRtcInterop.snapshot", cancellationToken, id, filename);

    public ValueTask MuteMic(string id, string participantId, CancellationToken cancellationToken = default) =>
        _jsRuntime.InvokeVoidAsync("TelnyxWebRtcInterop.muteMic", cancellationToken, id, participantId);

    public ValueTask MuteVideoParticipant(string id, string participantId, CancellationToken cancellationToken = default) =>
        _jsRuntime.InvokeVoidAsync("TelnyxWebRtcInterop.muteVideoParticipant", cancellationToken, id, participantId);

    public ValueTask Kick(string id, string participantId, CancellationToken cancellationToken = default) =>
        _jsRuntime.InvokeVoidAsync("TelnyxWebRtcInterop.kick", cancellationToken, id, participantId);

    public ValueTask VolumeUp(string id, string participantId, CancellationToken cancellationToken = default) =>
        _jsRuntime.InvokeVoidAsync("TelnyxWebRtcInterop.volumeUp", cancellationToken, id, participantId);

    public ValueTask VolumeDown(string id, string participantId, CancellationToken cancellationToken = default) =>
        _jsRuntime.InvokeVoidAsync("TelnyxWebRtcInterop.volumeDown", cancellationToken, id, participantId);

    public ValueTask<string?> GetCallStats(string id, CancellationToken cancellationToken = default) =>
        _jsRuntime.InvokeAsync<string?>("TelnyxWebRtcInterop.getCallStats", cancellationToken, id);

    public ValueTask SetAudioVolume(string id, double volume, CancellationToken cancellationToken = default) =>
        _jsRuntime.InvokeVoidAsync("TelnyxWebRtcInterop.setAudioVolume", cancellationToken, id, volume);

    public async ValueTask DisposeAsync()
    {
        await _resourceLoader.DisposeModule(_module);
        await _scriptInitializer.DisposeAsync();
    }
}